File: examples/python-unittest/steps/library.steps.py
from datetime import datetime

from library_example import FEE_PER_DAY, add_money, gbp, late_fee, may_borrow
from var import define_state


def to_date(raw):
    """June 6, 2026 → date(2026, 6, 6)."""
    return datetime.strptime(raw, "%B %d, %Y").date()


def format_date(d):
    """date(2026, 6, 6) → June 6, 2026 (no %-d — day padding flags are not portable)."""
    return f"{d:%B} {d.day}, {d.year}"


def to_money(raw):
    """£2.50 and 50p, both as GBP Money."""
    return gbp(float(raw[:-1]) / 100) if raw.endswith("p") else gbp(float(raw[1:]))


def format_money(m):
    """The inverse: mismatches render as £2.60 / 50p, not as a Money dump."""
    return f"{round(m.value * 100)}p" if m.value < 1 else f"£{m.value:.2f}"


stimulus, sensor = define_state(
    lambda: {"loans": (), "fee": gbp(0), "granted": False},
    {
        "date": {
            "regexp": r"[A-Z][a-z]+ \d{1,2}, \d{4}",
            "parse": to_date,
            "format": format_date,
        },
        "money": {
            # £2.50 and 50p, both as GBP Money. The amount is cucumber-expressions'
            # float regexp, minus the scientific notation.
            "regexp": r"£(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*|\d+p",
            "parse": to_money,
            "format": format_money,
        },
        # The emphasised run IS the parameter: the markers live in the pattern,
        # parse strips them, format restores them. Markup is notation, like £2.50.
        "title": {
            "regexp": r"\*[^*]+\*",
            "parse": lambda raw: raw[1:-1],
            "format": lambda t: f"*{t}*",
        },
    },
)


@stimulus("borrowed {title}, due back on {date}")
def _(state, title, due):
    return {"loans": (*state["loans"], {"title": title, "due": due})}


@stimulus("returns it on {date}")
def _(state, returned_on):
    fee = gbp(0)
    for loan in state["loans"]:
        fee = add_money(fee, late_fee(loan, returned_on))
    return {"fee": fee}


@sensor("owes a {money} late fee")
def _(state, expected):
    return state["fee"]


@sensor("{money} for each day overdue")
def _(state, expected):
    return FEE_PER_DAY


@stimulus("asks to borrow {title} on {date}")
def _(state, title, on):
    return {"granted": may_borrow(state["loans"], on)}


@sensor("the library refuses")
def _(state):
    if state["granted"]:
        raise AssertionError("expected the library to refuse")


@sensor("the library agrees")
def _(state):
    if not state["granted"]:
        raise AssertionError("expected the library to agree")

File: examples/ruby-rspec/steps/library.steps.rb
require "date"
require "oselvar/var"
require_relative "../lib/library"

# June 6, 2026 → Date; and the inverse (no day-padding flags — not portable).
to_date = ->(raw) { Date.strptime(raw, "%B %d, %Y") }
format_date = ->(d) { "#{d.strftime('%B')} #{d.day}, #{d.year}" }

# £2.50 and 50p, both as GBP Money.
to_money = ->(raw) { raw.end_with?("p") ? Library.gbp(raw[0...-1].to_f / 100) : Library.gbp(raw[1..].to_f) }
# The inverse: mismatches render as £2.60 / 50p, not as a Money dump.
format_money = ->(m) { m.value < 1 ? "#{(m.value * 100).round}p" : format("£%.2f", m.value) }

param, stimulus, sensor = steps { { loans: [], fee: Library.gbp(0), granted: false } }

param.("date", '[A-Z][a-z]+ \d{1,2}, \d{4}', parse: to_date, format: format_date)
# The amount is cucumber-expressions' float regexp, minus scientific notation.
param.("money", '£(?=.*\d.*)[-+]?\d*(?:\.(?=\d.*))?\d*|\d+p', parse: to_money, format: format_money)
# The emphasised run IS the parameter: the markers live in the pattern, parse
# strips them, format restores them. Markup is notation, like £2.50.
param.("title", '\*[^*]+\*', parse: ->(raw) { raw[1...-1] }, format: ->(t) { "*#{t}*" })

stimulus.("borrowed {title}, due back on {date}") do |state, title, due|
  { loans: state[:loans] + [{ title: title, due: due }] }
end

stimulus.("returns it on {date}") do |state, returned_on|
  fee = state[:loans].reduce(Library.gbp(0)) do |acc, loan|
    Library.add_money(acc, Library.late_fee(loan, returned_on))
  end
  { fee: fee }
end

sensor.("owes a {money} late fee") { |state, _expected| state[:fee] }

sensor.("{money} for each day overdue") { |_state, _expected| Library::FEE_PER_DAY }

stimulus.("asks to borrow {title} on {date}") do |state, _title, on|
  { granted: Library.may_borrow(state[:loans], on) }
end

sensor.("the library refuses") { |state| raise "expected the library to refuse" if state[:granted] }

sensor.("the library agrees") { |state| raise "expected the library to agree" unless state[:granted] }
